llvm/lib/ProfileData/Coverage/CoverageMapping.cpp

```diff
@@ -582,6 +583,87 @@ static unsigned getMaxBitmapSize(const CounterMappingContext &Ctx,
   return MaxBitmapID + (SizeInBits / CHAR_BIT);
 }
 
+/// This holds the DecisionRegion and MCDCBranch(es) under it.
+/// Also traverses Expansion(s).
+struct DecisionRow {
+  /// The subject
+  const CounterMappingRegion *DecisionRegion;
+
+  /// They are reflected from `DecisionRegion` for convenience.
+  LineColPair DecisionStartLoc;
+  LineColPair DecisionEndLoc;
+
+  /// This is passed to `MCDCRecordProcessor`, so this should be compatible to
+  /// `ArrayRef<const CounterMappingRegion *>`.
+  SmallVector<const CounterMappingRegion *, 6> Branches;
+
+  /// Each `ID` in `Branches` should be unique.
+  DenseSet<CounterMappingRegion::MCDCConditionID> IDs;
+
+  /// Relevant `Expansion`(s) should be caught to find expanded Branches.
+  SmallVector<const CounterMappingRegion *> Expansions;
+
+  DecisionRow(const CounterMappingRegion &Decision)
+      : DecisionRegion(&Decision), DecisionStartLoc(Decision.startLoc()),
+        DecisionEndLoc(Decision.endLoc()) {}
+
+  /// Determine whether `R` is included in `DecisionRegion`.
+  bool inDecisionRegion(const CounterMappingRegion &R) {
+    return (R.FileID == DecisionRegion->FileID &&
+            R.startLoc() >= DecisionStartLoc && R.endLoc() <= DecisionEndLoc);
+  }
+
+  /// Determin whether `R` is pointed by any of Expansions.
+  bool inExpansions(const CounterMappingRegion &R) {
+    return any_of(Expansions, [&](const auto &Expansion) {
+      return (Expansion->ExpandedFileID == R.FileID);
+    });
+  }
+
+  enum class UpdateResult {
+    NotFound = 0,
+    Updated,
+    Completed,
+  };
+
+  /// Add `Branch` into the Decision
+  UpdateResult updateBranch(const CounterMappingRegion &Branch) {
+    auto ID = Branch.MCDCParams.ID;
+    assert(ID > 0 && "MCDCBranch.ID should begin with 1");
+
+    if (!IDs.contains(ID) &&
```

Comment:
You can reduce indentation in this code by negating this `if`:

```
if (IDs.contains(ID) || (!inDecisionRegion(Branch) && !inExpansions(Branch))
  return UpdateResult::NotFound;
const auto NumConditions = DecisionRegion->MCDCParams.NumConditions;
assert(Branches.size() < NumConditions && "Have more than the max number of conditions?!");
/*...*/
return Branches.size() == NumConditions ? UpdateResult::Completed : UpdateResult::Updated;
```

It seems a little strange to me that `IDs.contains(ID)` means we return `UpdateResult::NotFound`. Why is that?